File: backend/src/pad.cpp
#include "pad.hpp"

using namespace Ats;
using namespace std;

Pad::Pad(uint stream,
	 uint chan,
	 uint pid,
	 std::string type,
         Glib::RefPtr<Gst::Bin> b,
	 Glib::RefPtr<Gst::Pad> p) {

    if ( type == "video" ) _t = Pad::Type::Video;
    else if ( type == "audio" ) _t = Pad::Type::Audio;
    else _t = Pad::Type::Unknown;

    _stream = stream;
    _channel = chan;
    _pid = pid;
    _bin = b;
    _tee = Gst::Tee::create();

    _bin->add(_tee);
    _tee->sync_state_with_parent();

    auto tee_sink = _tee->get_static_pad("sink");
    p->link(tee_sink);
    auto mid_pad = _tee->request_pad(_tee->get_pad_template("src_%u"));

    auto ghost = Gst::GhostPad::create(mid_pad);
    ghost->set_active();
    _pad       = (Glib::RefPtr<Gst::Pad>) ghost;
    _bin->add_pad(_pad);

    //_pad->signal_unlinked().connect ([this](const Glib::RefPtr<Gst::Pad> p){ if (p) _unlinked.emit(); });
}

Pad::~Pad() {
    _bin->remove_pad(_pad);
    if (_pad->is_ghost_pad ()) {
        auto gp = (Glib::RefPtr<Gst::GhostPad>*)(&_pad);
        _tee->remove_pad((*gp)->get_target());
    } else {
        _tee->remove_pad(_pad);
    }
}

std::shared_ptr<Pad>
Pad::copy() {
    auto p = std::shared_ptr<Pad> (new Pad());

    p->_stream   = _stream;
    p->_channel  = _channel;
    p->_pid      = _pid;
    p->_t        = _t;
    p->_tee      = _tee;
    p->_bin      = _bin;
    auto mid_pad = _tee->request_pad(_tee->get_pad_template("src_%u"));
    auto ghost   = Gst::GhostPad::create(mid_pad);
    ghost->set_active();
    p->_pad      = (Glib::RefPtr<Gst::Pad>) ghost;
    p->_bin->add_pad(p->_pad);

    return p;
}

File: backend/src/pad.rs
use gst;
use gst::prelude::*;

#[derive(Clone,Copy)]
pub enum Type {
    Video, Audio, Unknown //GraphVolume,
}

pub struct SrcPad {
    pub typ: Type,
    pub stream: String,
    pub channel: u32,
    pub pid: u32,

    bin: gst::Bin, // TODO remove
    tee: gst::Element,
    pub pad: gst::Pad,
}

impl SrcPad {
    pub fn new(stream: String, channel: u32, pid: u32, typ: &str, bin: gst::Bin, pad: &gst::Pad) -> SrcPad {
        let typ = match typ {
            "video" => Type::Video,
            "audio" => Type::Audio,
            _       => Type::Unknown,
        };

        let tee = gst::ElementFactory::make("tee", None).unwrap();
        bin.add(&tee).unwrap();
        tee.sync_state_with_parent().unwrap();

        let tee_sink = tee.get_static_pad("sink").unwrap();
        // TODO check
        let _ = pad.link(&tee_sink);

        let mid_pad = tee.get_request_pad("src_%u").unwrap();
        let ghost   = gst::GhostPad::new(None, &mid_pad).unwrap();
        ghost.set_active(true).unwrap();

        bin.add_pad(&ghost).unwrap();

        SrcPad { typ, stream, channel, pid, bin, tee, pad: ghost.upcast() }
    }
    
}

impl Clone for SrcPad {
    fn clone (&self) -> SrcPad {
        let mid_pad = self.tee.get_request_pad("src_%u").unwrap();
        let ghost   = gst::GhostPad::new(None, &mid_pad).unwrap();
        ghost.set_active(true).unwrap();

        SrcPad { typ: self.typ, stream: self.stream.clone(), channel: self.channel, pid: self.pid,
                 bin: self.bin.clone(), tee: self.tee.clone(), pad: ghost.upcast() }
    }
}

impl Drop for SrcPad {
    fn drop (&mut self) {
        self.bin.remove_pad(&self.pad).unwrap();
        if let Ok(p) = self.pad.clone().downcast::<gst::GhostPad>() {
            self.tee.remove_pad(&p.get_target().unwrap()).unwrap();
        } else {
            self.tee.remove_pad(&self.pad).unwrap();
        }
    }
}
